### PR TITLE
fix division by zero in normalizing

### PR DESCRIPTION
**Pull request by joonazan**
I don't know if you are still developing this as the latest change was in 2016. This is the only vector math library I found that installs on Elm 0.18.

If yes, you should update your version of elm-test and elm-format. I was not able to run the tests.

### CODE REVIEW

File: src/Vector2.elm [hunk 1]
```diff
@@ -243,14 +271,15 @@ distanceSquared v w =
...
 angle v w =
     let
         r =
             dot v w / (length v * length w)
     in
-        if r >= 1 then
-            0
-        else
-            acos r
+    if r >= 1 then
```
Comment (Zinggi) [inline src/Vector2.elm:282]: Is this correct? Wouldn' this give NaN for (0, 0) (0, 0)?
Shouldn't it be
```elm
let
    l = length v * length w
in
    if l == 0 then
        0
    else
        dot v w / l
```
?
If it is correct, then which version is faster?

**Comment by Zinggi** [comment]
Again, sorry for not replying to this for so long.
As for the code, see the above comment. The same goes for all `angle` functions

**Comment by Zinggi** [comment]
> you should update your version of elm-test and elm-format. I was not able to run the tests

I will do, thanks for letting me know.
I should probably also add doc tests, I have come to love those.

File: src/Vector2.elm (hunk 1 above)
Comment (joonazan) [reply src/Vector2.elm:282]: You are correct that that function could use as division by zero test as well.

I checked all the divisions in Vector2 and found that besides angle and normalize, project is also affected.

I would write project as `v dot (normalize w)`.

On second thought it is not clear what angle or normalize should return with zero vectors. At least in angle you can argue that zero is one valid answer for the angle between (0,0) and (0,0). In normalizing the behaviour that I implemented is most convenient for me, but it seems wrong to return a vector with length that is not one.

Comment (Zinggi) [reply src/Vector2.elm:282]: Well mathematically there is no right answer, so I think it makes sense to have a sane default.
But in any case, it should at least be documented.

> I would write project as v dot (normalize w)

This doesn't seem right. You can't get rid of the division.  
I think the best is to also just do a test for zero and in that case return 0.
Which, again, is mathematically nonsense, but it seems like a sane default.

**Comment by Zinggi** [comment]
Thinking a bit more about this, maybe it's ok to return NaN?

I think before actually doing this, we should do some research and see what other linear algebra libraries do.

Do they return 0?
Do they throw an exception?
Do they return NaN or infinity (e.g. they don't do anything special, they just let it happen)?

**Comment by joonazan** [comment]
I checked glsl. It has undefined behaviour when normalizing zero vectors.

Returning NaN is the correct approach, because code relying on a specific way of handling this is not nice.

On the other hand NaN spreads in a very ugly way, which is a thing that almost never happens in Elm. I'd argue that returning NaN inflicts too much pain on the affected programmer, as he will have no way to find the source corrupting all his numbers. (This is what happened to me at least.)

This leaves multiple options:
- `(0,0)` the only value that someone could actually want. Convenient, but does not encourage to use the library properly.
- `(1,0)` painful, but easy to find. Technically a valid return value, but will cause WTFs.
- `(666, 2414)` nope
- a random value: too much overhead

EDIT: with these options, my favorite is Debug.crash!

EDIT2: I just figured out that as these are floating point numbers, there can be fairly legitimate
calls of normalize on zero vectors that unlikely enough to not happen during testing. IMO that means that only (0,0) and (1,0) are acceptable. GLSL is different, as the undefined behaviour only messes up one fragment.

**Comment by Zinggi** [comment]
Well, after having looked at the other Elm vector math libraries,
I think leaving it as it is now is the right thing to do.
My reasoning is the following:

This library is basically a pure Elm version of the [elm-community linear algebra library](http://package.elm-lang.org/packages/elm-community/linear-algebra/latest).
It is meant as a fast vector math library, so any division by zero checking would add a slight overhead.

In practice, a user usually knows what kinds of vectors are fed to the problematic functions and in case some of these might be zero, the user can decide how to handle that before calling a problematic function.
I don't think deciding on any default value could make all users happy.

If you think that the library should take precautious measures to prevent these kind of bad situations, I would recommend you to use Ian Mackenzie's excellent [geometry](http://package.elm-lang.org/packages/opensolid/geometry/latest) library where it's version of `normalize` (called `direction` there) actually returns a `Maybe`.


>  I'd argue that returning NaN inflicts too much pain on the affected programmer, as he will have no way to find the source corrupting all his numbers. (This is what happened to me at least.)

I can sympathize with that, I was also bitten by this when I created a game with MonoGame. 
But I think the best thing that this library can do to combat that a little is to just document every problematic function, such that these errors are hopefully avoided.

**Comment by Zinggi** [comment]
Ok, so I updated the documentation to include NaN warnings for problematic functions (https://github.com/Zinggi/elm-webgl-math/commit/6ffbb5b958d7f5d659f8299d58047ce3e0291b99).   
I think that's the best course for this library.

This closes https://github.com/Zinggi/elm-webgl-math/issues/3
